pythonopensubtitles/opensubtitles.py

```diff
@@ -104,9 +109,99 @@ def detect_language(self):
         # array DetectLanguage( $token, array($text, $text, ...) )
         raise NotImplementedError
 
-    def download_subtitles(self):
-        # array DownloadSubtitles( $token, array($IDSubtitleFile, $IDSubtitleFile,...) )
-        raise NotImplementedError
+    def download_subtitles(self, subfile_ids, ext='srt',
+                           filenames=None, directory=None):
+        """
+        Returns a dictionary with max. 20 IDs of and paths to all
+        successfully downloaded subtitle files, otherwise None.
+
+        subfile_ids can be None/empty if a filenames dict is present;
+        otherwise, the file IDs used by OpenSubtitles are used as file
+        names, with .srt as default file extension.
+
+        Keep in mind that subtitle searches can contain multiple
+        results for the same file title (because several different
+        file uploads might reference the same file name).
+        So, if you use a filenames dict in which different file IDs
+        reference the exact same file name (i.e. IDs with the same value),
+        the number of eventually downloaded files will be smaller than the
+        overall number of file IDs submitted (there can only be one file
+        per file name, so any subsequent download attempts for an already
+        existing file will overwrite the previous version of said file).
+
+        :param subfile_ids: list of ids of subtitle files to download
+        :param filenames: optional dictionary with preferred file names;
+                          for keys use subtitle file IDs (as strings),
+                          for values use file names (incl. file extensions)
+        :param ext: one of: srt, sub, txt, ssa, smi, mpi, tmp;
+                    only applicable if only subtitle file IDs are provided
+        :param directory: path to preferred dir to which to write files
+        """
+        successful = {}
+        if filenames:
```

Comment:
I think it's better some duplication when you call the function than override one param with the other. Let me explain a little bit:

I think it's better to call the function with these params:
```
ids=[1,2,3]
override_filenames={
  1: 'batman.srt',
  2: 'floor_xii.srt',
  3: 'other_file.srt'
}
```

Than calling with:
```
ids=None
override_filenames={
  1: 'batman.srt',
  2: 'floor_xii.srt',
  3: 'other_file.srt'
}
```

What do you think?

Reply:
I'll reply to this later, when I'm on my main machine.

I think I did it like this so as to avoid having to deal with two instances of the same IDs - once in form of a list, once as dict keys - because it felt like unnecessary code duplication/extra work for the user when they are really only interested in matching the IDs to the file names (because they want to save those exact file names).

With your example, what exactly would the script do with the list of IDs if there is also a dict for overwriting the file names present? Would it always need/use the list for downloading the subtitles and only do the file name matching based on if an ID is present in the dict?

Reply:
Hmm, thinking about it some more, I was initially very unsure about dropping the list entirely - as in: ignoring it when a dict is present - I think *because* that list of IDs is part of the other functions.

So maybe my initial feeling was right and I should have kept it. (Basically for consistency.)

Reply:
I checked - yeah, I made the `ids` optional because I wanted to save the user (in this case: myself, obviously) time and effort, but I agree that it makes more sense to keep this function consistent with the rest.